cpmpy/expressions/globalconstraints.py

```diff
@@ -417,6 +417,33 @@ def get_bounds(self):
         bnds = [get_bounds(x) for x in self.args]
         return max(lb for lb,ub in bnds), max(ub for lb,ub in bnds)
 
+class Isin(GlobalConstraint):
+    """
+        The "isin" constraint, defining non-interval domains for an expression
+    """
+
+    def __init__(self, expr, arr):
+        assert not (is_boolexpr(expr) or any(is_boolexpr(a) for a in arr)), \
+            "The expressions in the Isin constraint should not be boolean"
+        assert len(arr) > 1, "The array given must contain more than 1 elements"
+        super().__init__("isin", [expr, arr], is_bool=True)
+
+    def decompose(self):
+        expr, arr = self.args
+        lb, ub = expr.get_bounds()
+        expressions = all(isinstance(a, Expression) for a in arr)
+        if expressions:
+            from .python_builtins import any
+            return [any(expr == a for a in arr)]
+        else:
+            return [expr != val for val in range(lb, ub + 1) if val not in arr]
```

Comment:
The idea here is to distinguish between a purely numeric array and an array of arbitrary expressions. What if we have a mixed array though? E.g., `[3, y, z]`.

Proposal: if `arr` is purely values, figure out the intervals it has, add the conjunction of disjunctions that `expr` is in those intervals. E.g., `IsIn(x,[3,5,6,7,8])` has two intervals - `3..3` and `5..8`. Then we get `x>=3 & (x=<3 | x>=5) & (x=<8)`. This can be way more efficient than iterating over individual values.